gym/envs/registration.py

```diff
@@ -513,36 +517,58 @@ def spec(self, path: str) -> EnvSpec:
-            message += " of the same environment already exists"
-            raise error.DeprecatedEnv(message)
-
         if self._ns is not None:
             if spec.namespace is not None:
                 logger.warn(
-                    f"Custom namespace '{spec.namespace}' is being overridden by namespace '{self._ns}'. "
-                    "If you are developing a plugin you shouldn't specify a namespace in `register` calls. "
-                    "The namespace is specified through the entry point key."
+                    f"Custom namespace `{spec.namespace}` is being overridden "
+                    f"by namespace `{self._ns}`. If you are developing a "
+                    "plugin you shouldn't specify a namespace in `register` "
+                    "calls. The namespace is specified through the "
+                    "entry point package metadata."
                 )
             # Replace namespace
             spec.namespace = self._ns
 
-        if spec.id in self.env_specs:
-            logger.warn(f"Overriding environment {id}")
-        self.env_specs[spec.id] = spec
+        try:
+            # Get all versions of this spec.
+            versions = self.env_specs.versions(spec.namespace, spec.name)
+
+            # We raise an error if the user is attempting to initialize an
+            # unversioned environment when a versioned one already exists.
+            latest_versioned_spec = max(
+                filter(lambda spec: isinstance(spec.version, int), versions),
+                key=lambda spec: cast(int, spec.version),
+                default=None,
+            )
+            unversioned_spec = next(
+                filter(lambda spec: spec.version is None, versions), None
+            )
+
+            # Trying to register an unversioned spec when versioned spec exists
+            if unversioned_spec and spec.version is not None:
+                message = (
+                    "Can't register the versioned environment "
+                    f"`{spec.id}` when the unversioned environment "
+                    f"`{unversioned_spec.id}` of the same name already exists."
+                )
+                raise error.RegistrationError(message)
+            elif latest_versioned_spec and spec.version is None:
+                message = (
+                    f"Can't register the unversioned environment `{spec.id}` "
+                    f"when version `{latest_versioned_spec.version}` "
+                    "of the same name already exists. Note: the default "
+                    "behavior is that the `gym.make` with the unversioned "
+                    "environment will return the latest versioned environment."
+                )
+                raise error.RegistrationError(message)
+        # We might not find this namespace or name in which case
+        # we should continue to register the environment.
+        except (error.NamespaceNotFound, error.NameNotFound):
```

Comment:
Not sure how I feel about this bare except but I think it's fine for our purposes. If we can't find a namespace or name then we most definitely want to continue to register the environment.

Reply:
Well it works, and I don't have a better suggestion either :sweat_smile: